### ISSUE TITLE
Not using all cores

### ISSUE THREAD

**Issue by DiegoCorrea** [joblib/joblib#1147]
Hi, can you help me in an issue?
I'm using python 3.7 and joblib 0.17 on Red Hat and Ubuntu.

So, I want to use all my 240 cores and to this a have 5,17 TB of ram, core and memory is not my problem. But when I try to use all this cores the with n_jobs=N_CORES or n_jobs=-2 or n_jobs=-1, the joblib not run all my jobs. In total I will run 243 jobs.  
That is the snippet code:
```
cv = get_cv(self.cv)

delayed_list = (
    delayed(fit_and_score)(self.algo_class(**params), trainset,
                           testset, self.measures,
                           self.return_train_measures)
    for params, (trainset, testset) in product(self.param_combinations,
                                               cv.split(data))
)
out = Parallel(n_jobs=self.n_jobs,
               pre_dispatch=self.pre_dispatch,
               verbose=self.joblib_verbose)(delayed_list)

(test_measures_dicts,
 train_measures_dicts,
 fit_times,
 test_times) = zip(*out)
```

This code belongs to the Surprise lib and I want to run the Grid Search to SVDpp/NMF/USERKNN. Here is the original code: https://github.com/NicolasHug/Surprise/blob/master/surprise/model_selection/search.py  

So, all images will show you that the joblib does not use all cores in the machine.  

![svdpp](https://user-images.githubusercontent.com/3698831/103588542-2f5d4700-4ec8-11eb-9c52-1d427180c065.png)
![userknn](https://user-images.githubusercontent.com/3698831/103588553-3421fb00-4ec8-11eb-8440-b5ce275aa3a2.png)
![nmf](https://user-images.githubusercontent.com/3698831/103588619-59af0480-4ec8-11eb-96cf-b4eb97a257ba.png)

Some of job have more than 450 minutes of run and just 3 cores used.  

Sorry if I ask something out of scope. But I try to solve this problem and nothing happens.

**Comment by tomMoral** [comment]
Hi,

Could you check if running this code, you have a large number (~240) line in your top with `python3.7` (they might be idle)? Just to be sure you indeed have 240 jobs created. If not, this would be an issue on `joblib` side.

Another possibility is that the issue here is one of load balancing. You might have jobs that are very slow while the other are very fast, so you only observe the one that take a lot of time.

A potential cause for this would be our oversubscription mitigation mechanism. When you use `n_jobs=-1/-2`, `joblib` makes sure that each job will only use 1 CPU to avoid over-subscription. Maybe you would get better performance using `N_CORES // 4`. Indeed, in this case, each job can use up to 4 CPUs for `numpy/BLAS` operation and this can reduce the runtime of the longer jobs. Could you try an report back if using this reduce the running time and make use of more CPUs?

**Comment by DiegoCorrea** [comment]
Hi Tom, thanks for read and answer me,  
About you comment: 

> Could you check if running this code, you have a large number (~240) line in your top with python3.7 (they might be idle)? Just to be sure you indeed have 240 jobs created. If not, this would be an issue on joblib side.

How I know if I had all 240 jobs created?   

> Another possibility is that the issue here is one of load balancing. You might have jobs that are very slow while the other are very fast, so you only observe the one that take a lot of time.

My dataset has more than 10M training line and it get sometime to finish the job. I try to make a Grid Search in 2 ML algo, SVD++(very slow algo) and NMF(8 params to search, if I try 3 possibilities is 3^8 = 6561 jobs to create with the same dataset). So, they use the same time to finish.

> A potential cause for this would be our oversubscription mitigation mechanism. When you use n_jobs=-1/-2, joblib makes sure that each job will only use 1 CPU to avoid over-subscription. Maybe you would get better performance using N_CORES // 4. Indeed, in this case, each job can use up to 4 CPUs for numpy/BLAS operation and this can reduce the runtime of the longer jobs. Could you try an report back if using this reduce the running time and make use of more CPUs?

I try `n_jobs=N_CORES` and `n_jobs=N_CORES//4` as you said.  
OBS:To this test I get a machine with 24 cores and 62 GB ram. Is more easy allocate this machine to test.  And I can get this machine for 20 minutes to test.

With `n_jobs=N_CORES` I get this:
```
07-Jan-21 17:39:01 - __main__ - INFO - > N Cores: 24 RAM: 62.72GB
07-Jan-21 17:39:01 - __main__ - INFO - start at 17:39:01
[Parallel(n_jobs=24)]: Using backend LokyBackend with 24 concurrent workers.
/diego.silva/.conda/envs/venv/lib/python3.7/site-packages/joblib/externals/loky/process_executor.py:691: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  "timeout or by a memory leak.", UserWarning
[Parallel(n_jobs=24)]: Done   1 tasks      | elapsed:  7.8min
[Parallel(n_jobs=24)]: Done   2 tasks      | elapsed: 10.1min
[Parallel(n_jobs=24)]: Done   3 tasks      | elapsed: 12.5min
[Parallel(n_jobs=24)]: Done   4 tasks      | elapsed: 14.4min
```
![nvidia_dev_nmf_njob_by_4](https://user-images.githubusercontent.com/3698831/103946067-ecdc7a00-5114-11eb-9797-606f1e5295e9.png)

I set to using all 24 cores but just 4 jobs are work in parallel.

With `n_jobs=N_CORES//4` I get this:
OBS: 24/4=6...
```
07-Jan-21 20:11:38 - __main__ - INFO - > N Cores: 24 RAM: 62.72GB
07-Jan-21 20:11:38 - __main__ - INFO - start at 20:11:38
[Parallel(n_jobs=6)]: Using backend LokyBackend with 6 concurrent workers.
/diego.silva/.conda/envs/venv/lib/python3.7/site-packages/joblib/externals/loky/process_executor.py:691: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  "timeout or by a memory leak.", UserWarning
[Parallel(n_jobs=6)]: Done   1 tasks      | elapsed:  8.2min
[Parallel(n_jobs=6)]: Done   2 tasks      | elapsed: 10.4min
[Parallel(n_jobs=6)]: Done   3 tasks      | elapsed: 12.7min
[Parallel(n_jobs=6)]: Done   4 tasks      | elapsed: 15.3min
```
![nmf_dev_6](https://user-images.githubusercontent.com/3698831/103956200-ccb5b680-5126-11eb-8145-4a11712fc238.png)

As you see, the same thing happens when I use all cores and n_cores//4.
